pkg/util/pki/asn1_util_test.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2023 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pki
+
+import (
+	"encoding/asn1"
+	"errors"
+	"reflect"
+	"testing"
+)
+
+func TestParseObjectIdentifier(t *testing.T) {
+	testCases := []struct {
+		oidString   string
+		expectedOid asn1.ObjectIdentifier
+		expectedErr error
+	}{
+		{
+			oidString:   "1.2.3.4.5",
+			expectedOid: asn1.ObjectIdentifier{1, 2, 3, 4, 5},
+			expectedErr: nil,
+		},
+		{
+			oidString:   "1.2.840.113549.1.1.1",
+			expectedOid: asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 1, 1},
+			expectedErr: nil,
+		},
+		{
+			oidString:   "1.3.6.1.4.1.311.60.2.1.3",
+			expectedOid: asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 311, 60, 2, 1, 3},
+			expectedErr: nil,
+		},
+		{
+			oidString:   ".",
+			expectedOid: nil,
+			expectedErr: errors.New("strconv.Atoi: parsing \"\": invalid syntax"),
+		},
+		{
+			oidString:   ".555",
+			expectedOid: nil,
+			expectedErr: errors.New("strconv.Atoi: parsing \"\": invalid syntax"),
+		},
+		{
+			oidString:   "555.",
+			expectedOid: nil,
+			expectedErr: errors.New("strconv.Atoi: parsing \"\": invalid syntax"),
+		},
+		{
+			oidString:   "test.5",
+			expectedOid: nil,
+			expectedErr: errors.New("strconv.Atoi: parsing \"test\": invalid syntax"),
+		},
+	}
+
+	for _, tc := range testCases {
+		oid, err := ParseObjectIdentifier(tc.oidString)
+		if err != nil {
+			if tc.expectedErr == nil {
+				t.Errorf("Unexpected error: %v", err)
+			} else if err.Error() != tc.expectedErr.Error() {
+				t.Errorf("Expected error: %v, got: %v", tc.expectedErr, err)
+			}
+		} else if !oid.Equal(tc.expectedOid) {
+			t.Errorf("Expected OID: %v, got: %v", tc.expectedOid, oid)
+		}
+	}
+}
+
+func TestMarshalAndUnmarshalUniversalValue(t *testing.T) {
+	testCases := []struct {
+		name                string
+		uv                  UniversalValue
+		raw                 asn1.RawValue
+		overrideRoundtripUv *UniversalValue
+	}{
+		{
+			name: "Test with IA5String",
```

Comment:
suggestion (non-blocking): For tests with names like this, I like to keep the name as a map key so the struct is only test data. Also means your tests are likely to be run in random order which is what I usually want!

It's definitely not important though, feel free to ignore.

```suggestion
	testCases := map[string]struct {
		uv                  UniversalValue
		raw                 asn1.RawValue
		overrideRoundtripUv *UniversalValue
	}{
		"Test with IA5String": {
```

Reply:
done